drivers: i2c: mspm0: fix target_un/register deadlock

hansbinderup opened the pull request:
See each commit for details :)

hansbinderup committed:
drivers: i2c: mspm0: use hex format for target addresses

This is the format used in the DTS why it's nice to have those aligned.

Signed-off-by: Hans Binderup <habi@bang-olufsen.dk>

hansbinderup committed:
drivers: i2c: mspm0: clang-format comment

The comment in this commit has not been properly formatted why it's
always part of the diff when working on this file. Ensure that the
formatting is aligned with the rest of the driver.

Signed-off-by: Hans Binderup <habi@bang-olufsen.dk>

hansbinderup commented on drivers/i2c/i2c_mspm0g3xxx.c:
```diff
@@ -414,16 +414,13 @@ static int i2c_mspm0g3xxx_target_register(const struct device *dev,
 		return -ENOTSUP;
 	}
 
-	k_sem_take(&data->i2c_busy_sem, K_FOREVER);
```
I've looked into other I2C drivers and they often don't lock un/register callbacks. I'm not sure how I feel about this.
I feel that it's becoming racy.
A solution would be to irq_lock but I'm also not sure how much I like that solution.

Svendsen991 commented on drivers/i2c/i2c_mspm0g3xxx.c:
```diff
@@ -448,12 +445,12 @@ static int i2c_mspm0g3xxx_target_register(const struct device *dev,
 		return 0;
 	}
 
+	k_sem_take(&data->i2c_busy_sem, K_FOREVER);
```
Do you know which one of the functions above is causing the deadlock?

hansbinderup replied:
Both - we're calling them from the `i2c_target_callbacks` (the target driver that un/register other i2c_target drivers).

Emil-Juhl commented on drivers/i2c/i2c_mspm0g3xxx.c:
```diff
@@ -525,16 +519,16 @@ static int i2c_mspm0g3xxx_target_unregister(const struct device *dev,
 		return 0;
 	}
 
+	k_sem_take(&data->i2c_busy_sem, K_FOREVER);
```
Just so I follow 100%, this will never be called in an isr context? If that's not guaranteed, the semaphore needs to either be replaced, or always taken with `K_NO_WAIT` (api states this)

hansbinderup replied:
We can't guarantee that. 
Had rubber duck session with @Svendsen991. I think that I have a better way to solve this (hint: we have a target serialization mechanism implemented hehe) :)

Emil-Juhl replied:
Yeah, I was thinking if there is no guarantee, we should use the workqueue which is already there. Super. I will take another look later today. The diff needs some closer inspection this time around ;)

hansbinderup replied:
Exactly! This part of the driver always crashed when enabling assertions so I wanted to fix that for a while :laughing: it's from the "upstream TI driver". But they haven't added the msgq (yet at least), so I think it's fine that we deviate as long as we're utilizing the mechanisms that are already deviating :)

hansbinderup commented:
@Emil-Juhl @Svendsen991 after looking into the un/register API I noticed that it does in fact not specify whether it's supposed to be a/sync. We could add some sync mechanism. But in reality we have limited amount of targets that can be registered (for mspm0 this is 2 targets), why you are forced to keep track of this amount anyways.

I think that this is the simplest solution; it's fully thread and race safe and it's synchronized by the same mechanism as the other target handles. 

```
shell test:
mspm0:-$ devices i2c register emulated-pca954x@48 1
Succesfully registered emulated-pca954x@48
[00:12:53.338,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_register: Register primary target on addr=0x48
mspm0:-$ devices i2c register emulated-pca953x@77 1
Succesfully registered emulated-pca953x@77
[00:12:57.581,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_register: Register alternate target on addr=0x77
mspm0:-$ devices i2c register emulated-pca953x@77 0
Succesfully unregistered emulated-pca953x@77
[00:13:03.191,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_unregister: Unregistering alternate on addr=0x77
mspm0:-$ devices i2c register emulated-pca954x@48 0
Succesfully unregistered emulated-pca954x@48
[00:13:06.802,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_unregister: Unregistering primary on addr=0x48
[00:13:06.803,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_unregister: Enabling controller mode as no targets are registered


Linux->mspm0 test (using proprietary B&O drivers):
1000.0.0.39035081@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x01
[00:13:54.203,000] <dbg> emul_pca954x: emul_pca954x_apply_mux_config: apply mux_config=1, mask=1
[00:13:54.203,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_register: Register alternate target on addr=0x77
1000.0.0.39035081@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x00
[00:14:12.834,000] <dbg> emul_pca954x: emul_pca954x_apply_mux_config: apply mux_config=0, mask=1
[00:14:12.834,000] <dbg> i2c_mspm0g3xxx: i2c_mspm0g3xxx_target_work_unregister: Unregistering alternate on addr=0x77
```

Emil-Juhl commented:
> @Emil-Juhl @Svendsen991 after looking into the un/register API I noticed that it does in fact not specify whether it's supposed to be a/sync. We could add some sync mechanism.
>
Yeah, but I agree with you in terms of not doing this unless really necessary. If the API doesn't specify requirements, then I guess the driver just have to be well behaved on its own (i.e. if you call `register` and then do some i2c ops, the register operation must complete before the i2c operation).

> But in reality we have limited amount of targets that can be registered (for mspm0 this is 2 targets), why you are forced to keep track of this amount anyways.
> 
Didn't quite catch the relevance with regards to sync or async api, but yeah, there'll have to be some book keeping for that.
> I think that this is the simplest solution; it's fully thread and race safe and it's synchronized by the same mechanism as the other target handles.
> 
Agreed, so long as it is good enough.
One question in my mind is whether this affects the flow of events to the "outside" world. As an example, the pca954x emulation could be required to have completed the i2c muxing before it ACKs to the controller, such that any i2c communication performed immediately after the ACK will be forwarded to the selected channel.
Will handling these register and unregister on the workqueue change the effective behavior for this case? (just as an imagined example, I don't actually _know_ the pca954x chips that well :) ).
And here I am thinking the behavioral aspect in terms of making sure that the selected channel won't miss data if it arrives too soon. The system latency is of course a bit different, but that's most likely not an issue - or otherwise it probably was without these changes anyway.

hansbinderup commented:
> One question in my mind is whether this affects the flow of events to the "outside" world. As an example, the pca954x emulation could be required to have completed the i2c muxing before it ACKs to the controller, such that any i2c communication performed immediately after the ACK will be forwarded to the selected channel.
Will handling these register and unregister on the workqueue change the effective behavior for this case? (just as an imagined example, I don't actually know the pca954x chips that well :) ).

I just did some measurements and it seems that the I2C hardware controls acking before we handle the un/register callbacks.

I've added control of acking (will be separate commit) which should solve this issue. TI also added this to their "upstream" so it should be safe to do.

With ack override:
```
1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x01
$6 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                   
$7 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                 
$8 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                 
$9 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_NO_INT, tconfig = 0x20200298 <emul_pca953x_0_data>, msgq_type = TARGET_MSGQ_REGISTER}                         
$10 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                   

1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x00
$11 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                  
$12 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                
$13 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                
$14 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_NO_INT, tconfig = 0x20200298 <emul_pca953x_0_data>, msgq_type = TARGET_MSGQ_UNREGISTER}                      
$15 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA} 
```

without ack override:
```
1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x01
$6 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                   
$7 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                 
$8 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                    
$9 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_NO_INT, tconfig = 0x20200298 <emul_pca953x_0_data>, msgq_type = TARGET_MSGQ_REGISTER}              

1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x00   
$10 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                  
$11 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                
$12 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>, msgq_type = TARGET_MSGQ_I2C_DATA}                   
$13 = {data = 0x20200420 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_NO_INT, tconfig = 0x20200298 <emul_pca953x_0_data>, msgq_type = TARGET_MSGQ_UNREGISTER} 
```

hansbinderup committed:
drivers: i2c: mspm0: add ack override handling

In the hunt of race conditions I did some low level traces following the
IRQ and msgq order.
It turned out that a normal case scenario would look like:

IRQ START -> IRQ RX/TX -> IRQ STOP -> HANDLE RX/TX

This is an issue as we might rely on timing happing in the "RX/TX"
process. With this commit the flow is now:

IRQ START -> IRQ RX/TX -> HANDLE RX/TX -> ACK/NACK -> IRQ STOP

which means that we can ensure the callbacks handled in RX/TX are
executed before we ack/nack and then stop.

Signed-off-by: Hans Binderup <habi@bang-olufsen.dk>

hansbinderup committed:
drivers: i2c: mspm0: fix locking in target_un/register

The target register methods had a few issues.

They would enforce locking which is a problem if:
- called from target msgq context (lock already taken -> deadlock)
- called from isr (will crash)

This commit takes the context into account and ensures that it's safe to
call this method from isr and the msgq context as well as any normal
context.

Signed-off-by: Hans Binderup <habi@bang-olufsen.dk>

hansbinderup commented:
Decided to take a step back.
I didn't like the fact that the un/register methods were no longer able to return a status and also that we were not longer able to wait for it. It would implicitly happen due to the target_work msgq priority why it worked. But these are configurable and therefore not a stable solution.

Instead I've decided to fix the problem more locally; deal with the issues in the methods themselves.
Now we handle the lock based on context. Should give us explicitness, return status and guaranteed execution before returning.

Dump from GDB to show that call order is still guaranteed:

```
1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x01
$30 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>}
$31 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
$32 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
Breakpoint 2, i2c_mspm0g3xxx_target_register (dev=0xf798 <__device_dts_ord_47>, target_config=0x20200298 <emul_pca953x_0_data>) at /workspaces/map-mcus
/zephyr/drivers/i2c/i2c_mspm0g3xxx.c:432
432             const struct i2c_mspm0g3xxx_config *config = dev->config;
$33 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>}

1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x10 0x00
$34 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>}
$35 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
$36 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
Breakpoint 3, i2c_mspm0g3xxx_target_unregister (dev=0xf798 <__device_dts_ord_47>, target_config=0x20200298 <emul_pca953x_0_data>) at /workspaces/map-mc
us/zephyr/drivers/i2c/i2c_mspm0g3xxx.c:546
546             const struct i2c_mspm0g3xxx_config *config = dev->config;
$37 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>}
```

and with the ack override we also introduced some improvements. We now nack when writing invalid registers:
```
1000.0.0.41454275@imx8mm-wb15:~# i2cset -f -y 11 0x48 0x06 0x92
Error: Write failed

GDB:
$38 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_START, tconfig = 0x20200268 <emul_pca954x_0_data>}
$39 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
$40 = 0
$41 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_RX_DONE, tconfig = 0x20200268 <emul_pca954x_0_data>}
$42 = -13
$43 = {data = 0x20200424 <i2c_mspm0g3xxx_data_0>, i2c_iidx = DL_I2C_IIDX_TARGET_STOP, tconfig = 0x20200268 <emul_pca954x_0_data>}

```
